bundle/edu.gemini.ictd/src/main/scala/edu/gemini/ictd/dao/FeatureTablesDao.scala

```diff
@@ -0,0 +1,111 @@
+package edu.gemini.ictd.dao
+
+import edu.gemini.ictd._
+import edu.gemini.pot.sp.Instrument
+import edu.gemini.spModel.ictd.Availability
+
+import doobie.imports._
+
+import scala.collection.immutable.TreeMap
+
+import scalaz._
+import scalaz.Scalaz._
+
+import shapeless.tag.@@
+import shapeless.{ HNil, tag }
+
+/** DAO for standard instrument features like filters and gratings. */
+object FeatureTablesDao {
+
+  final case class Feature(
+    name:       String,
+    instrument: Instrument,
+    location:   Location
+  )
+
+  // Tag the various kinds of Feature to distinguish them.  They are otherwise
+  // identical triplets of (name, Instrument, Location).
+  type Circular   = Feature @@ CircularTag
+  type Filter     = Feature @@ FilterTag
+  type Grating    = Feature @@ GratingTag
+  type Ifu        = Feature @@ IfuTag
+  type Longslit   = Feature @@ LongslitTag
+  type NSLongslit = Feature @@ NSLongslitTag
+
+  val select: ConnectionIO[Map[Instrument, FeatureTables]] =
+    for {
+      cs <- Statements.selectCircular.list
+      fs <- Statements.selectFilter.list
+      gs <- Statements.selectGrating.list
+      is <- Statements.selectIfu.list
+      ls <- Statements.selectLongslit.list
+      ns <- Statements.selectNsLongslit.list
+    } yield featureTables(cs, fs, gs, is, ls, ns)
+
+  private def featureTables(
+    cs: List[Circular],
+    fs: List[Filter],
+    gs: List[Grating],
+    is: List[Ifu],
+    ls: List[Longslit],
+    ns: List[NSLongslit]
+  ): Map[Instrument, FeatureTables] = {
+
+    def toAvailabilityTable[T](features: List[Feature @@ T]): AvailabilityTable @@ T =
+      tag[T][AvailabilityTable](features.map(f => (f.name, f.location.availability)).toMap)
+
+    def group[T](features: List[Feature @@ T]): Map[Instrument, AvailabilityTable @@ T] =
+      features.groupBy(_.instrument)
+              .mapValues(toAvailabilityTable)
+              .withDefaultValue(tag[T][AvailabilityTable](Map.empty))
```

Comment:
When you use `.withDefaultValue` it breaks a lot of map invariants, so it's safer in these cases to widen to a function type. Would things work if the return type here were `Instrument => AvailabilityTable @@ T` ?

Same comment with the use of `withDefault` elsewhere.